File: 0946-validate-stack-sequences/0946-validate-stack-sequences.cpp
class Solution {
public:
    bool validateStackSequences(vector<int>& v1, vector<int>& v2) {
        stack<int> st;
        int j = 0;
        for(int i = 0; i<v1.size() and j < v2.size(); i++){
            st.push(v1[i]);
            while(st.size() and st.top() == v2[j]){
                st.pop();
                j++;
            }
        }
        if(st.size() == 0)return true;
        while(st.size() and j < v2.size()){
            if(st.top() == v2[j]){
                st.pop();
                j++;
            }
            else{
                return false;
            }
        }
        return true;
    }
};

File: 0946-validate-stack-sequences/0946-validate-stack-sequences.java

class Solution {
    public boolean validateStackSequences(int[] pushed, int[] popped) {
        Stack<Integer> st = new Stack<>();
        int n = pushed.length;
        int j = 0;
        for(int i = 0; i<n; i++){
            if(st.isEmpty() || st.peek()!=popped[j]) st.push(pushed[i]);
            while(!st.isEmpty() && st.peek() == popped[j]){
                st.pop();
                j++;
            }
        }
        if(j!=n-1){
            while(j<n){
                if(st.peek()==popped[j]){
                    st.pop();
                    j++;
                }else{
                    return false;
                }
            }
        }
        return true;
    }
}